### ISSUE TITLE
[Middleware] Replace queryVal or context along the way

### ISSUE THREAD

**Issue by luisobo** [sangria-graphql/sangria#88]
Hey there,

I'm implementing authorization and I have the current scenario: the request gets a token in a header, I can decode this token locally for authentication but if any of the fields that were requested required to check the roles I have a query an external service to get the roles of the user. Currently I have these options:

- in the before query make that api call so roles info is available to every `beforeField` call. Problem here is that I'd paying the cost of that api call even for the fields that do not require authorization. This is a no go
- Making the api call for every field that requires authorization. Suboptimal too, I'd be making the same api call multiple times per query. I don't have many fields that require this but I much rather no go with this.

The ideal scenario for me would be to do that basic and local authentication in the beforeQuery and in the first field, and only the first, that has a Permission tag, make the external api call and replace the current user object (without role info) with the one I just got from the external service, the one with roles. That way, if any other field needs roles, it does not have to make the external api call.

One way to achieve this is by retuning a queryVal in beforeField, so that value becomes the new queryVal.

I'm not sure how this would play with parallel queries to be honest.

That or any other solution that satisfies those limitations would be very much appreciated. In any case, thanks for you time!

**Comment by luisobo** [comment]
Forgot to mention, maybe an option would be to collect the FieldTags at the beforeQuery and make the decision there, that be, actually, better (for this use case anyway)

**Comment by OlegIlyenko** [comment]
Thanks for the detailed description! I think I got the idea, I also need to think how to make it possible or whether it's already possible with the current feature set. Maybe we can introduce something like query visitor (which is conceptually similar to query complexity analysis which is just released) which will run on the query tree and gather the meta-information. This way you can collect all of the permissions before query even executed.

**Comment by OlegIlyenko** [comment]
I played a bit with some ideas... I think that `beforeField` would be wrong place for this kind of logic: it's already too late since query execution is already started. As you mentioned, the parallel execution is a big concern here as well. The tag collection is a very good idea I think, but `Middleware` is not very suitable for this.

I generalized the complexity analysis and rewritten it in terns of simple reduce functions. This new mechanism (`QueryReducer`) can also be used for other things like `FieldTag` collection. You can use it like this:

```scala
val permissionFetcher = QueryReducer.collectTags[MyContext, String] {
  case Permission(name) ⇒ name
}  { (permissionNames, ctx) ⇒
  val permissions = ctx.myPermissionService.fetchPermissions(permissionNames)
  
  Right(ctx.copy(permissions = permissions))

  // or if `permissions` is `Future[T]`
  // Left(permissions map (p ⇒ ctx.copy(permissions = p)))
}

Executor.execute(schema, query,
          userContext = new MyContext(...),
          queryReducers = permissionFetcher :: Nil)
```

`permissionFetcher` will go though the normalized query tree and will collect all of the permissions. Then the second function would be called with the collected results. it allows you to transform/enhance `Ctx`. All these things happen before query is executed. What do you think, will this help in your case? 

It's not released yet, but I published the snapshot:

```scala
resolvers += "Sonatype snapshots" at "https://oss.sonatype.org/content/repositories/snapshots/"
libraryDependencies += "org.sangria-graphql" %% "sangria" % "0.4.3-SNAPSHOT"
```

I would appreciate you if you could try it out and share your feedback and any thoughts concerning this new mechanism.

**Comment by luisobo** [comment]
Like a charm! 

Some thoughts:
I'm using a future to fetch permissions. Using an Either to return either a future or a value seems a little arbitrary and may not be as extensible. I really like how Actions are modeled and you can use a future, a value, a deferred, etc. Maybe something like that could be better.

It may be because of the future, but the updated context is not available in the middleware on the `beforeQuery`. Not a big deal, I can extract the user in the `beforeField` but that may not work in other cases. If it is because the future is not completed by the time of `beforeQuery`, consider waiting for the future?

Thanks for this change!

**Comment by OlegIlyenko** [comment]
Thanks a lot for the feedback! 

> Using an Either to return either a future or a value seems a little arbitrary and may not be as extensible.

I agree. I guess it was too quick decision in my side :) I will think about a better solution. Maybe something like `Action` or maybe I can just allow a subset of possible `Action`s subtypes.

> updated context is not available in the middleware on the beforeQuery

It should be available, have you tried this one (it very recent change):

https://github.com/sangria-graphql/sangria/blob/1cd155c34d195d5bf311573786570c11d9add812/src/main/scala/sangria/execution/middleware.scala#L34

**Comment by luisobo** [comment]
Works! Thanks

Another thing, I'm having trouble type checking the type of the my own context inside the middleware and query reducer. Could these types be parameterize to the context type?

The middleware ignores the type:
`def beforeQuery(context: MiddlewareQueryContext[_, _])`

and the QueryReducer parameterize each method but not the type itself
` def reduceField[Ctx, Val]`

Then there is the option that I may be missing something.

Thanks!

**Comment by OlegIlyenko** [comment]
Thanks for the feedback, it was very helpful! 

As I started to implement the `Middleware` it had a type parameter, but after several frustrating attempts to make it type-check in the `Executor` i gave up (too easily :) ). When I started with `QueryReducer`, I already had an assumption that it will not be possible to make it type-check, that's why I left out the type parameter for `Ctx` on the `QueryReducer`.

 Thanks for giving motivation to revisit this area! Turns out that with fair amount of co- and contravariance it can be possible. I published another snapshot with parametrized `Middleware` and `QueryReducer`. It would be great if you could check it out.

BTW, I also get rid of this `Either` and replaced it with proper `Action` (in fact a limited subset of actions :`ReduceAction`). So now you should be able to just return a `Future[Ctx]`, `Try[Ctx]` or `Ctx` out of the `reduceCtx` and it would be treated as an `Action`.

**Comment by luisobo** [comment]
Done, done and done. That's amazing. Everything works great, thank you!

No pressure but, do you know when  this is gonna hit the streets? Just curious cause I don't wanna ship this pointing to a snapshot. 

Thanks again!

**Comment by OlegIlyenko** [comment]
Shipping with snapshot dependencies is not good :) I just released v0.4.3 with all these changes :ship:
